gatsby/content/docs/e2ee_cross_signing.mdx

```diff
@@ -0,0 +1,920 @@
+  if (info.algorithm != 'm.pbkdf2') {
+    throw 'Unknown algorithm';
+  }
+  var generator = PBKDF2(hashAlgorithm: sha512);
+  return Uint8List.fromList(generator.generateKey(password, info.salt,
+      info.iterations, info.bits != null ? info.bits / 8 : 32));
+}
+```
+
+Important: Some PBKDF2 algorithms take as length the length in bytes (like here), while others take
+the length in bits. Divide by 8 appropriately, and only when needed.
+
+#### Validating the generated key
+To know if the right keyfile was supplied or if the user entered the right password, we need to verify
+the resulting keys validity. For that, the other information of the keys object in account data is
+relevant:
+
+```json
+{
+  "algorithm": "m.secret_storage.v1.aes-hmac-sha2", // some old accounts might use a different algorithm, which is not covered in this guide
+  "iv": "<base64 encoded iv>",
+  "mac": "<bae64 encoded mac>"
+}
+```
+
+To verify the key, encrypt 32 bytes of 0 with a blank name and the IV of the key information. Compare
+the resulting macs with each other. Be sure to compare the *unpadded* base64-encoded macs, so strip
+all the trailing `=`'s! The encryption itself will be outlined below.
+
+```dart
+bool checkKey(Uint8List key, AccountData keyData) {
+  var info = keyData.content;
+  if (info['algorithm'] == 'm.secret_storage.v1.aes-hmac-sha2') {
+    if ((info['mac'] is String) && (info['iv'] is String)) {
+      // ZERO_STR are 32 bytes of zero. We encrypt with our generated key, a blank name and the iv of the event
+      var encrypted = encryptAes(ZERO_STR, key, '', info['iv']);
+      // stripping all the trailing = of the macs prior comparing
+      return info['mac'].replaceAll(RegExp(r'=+$'), '') ==
+          encrypted.mac.replaceAll(RegExp(r'=+$'), '');
+    } else {
+      // no real information about the key, assume it is valid
+      return true;
+    }
+  } else {
+    throw 'Unknown Algorithm';
+  }
+}
+```
+
+### Encrypting and decrypting
+Both encrypting and decrypting is done with AES-CTR. Out of the secret key and the name of the secret
+using [HKDF](https://en.wikipedia.org/wiki/HKDF) with SHA-256 both the actual AES key and the key for
+the MAC integrity is generated. IVs are randomly generated 16 bytes, with bit 63 set to 0. In the
+account data events of the secrets, the mac, the IV and the ciphertext are all base64 encoded.
+
+Note: Encrypting and decrypting both uses *unpadded* AES-CTR. If your library does not have an option
+for unpadded AES-CTR en/decryption, simply append prior en/decryption 0's to your data to fill up the
+input to a full number of blocks. One block here is typically 32? bytes in size. After en/decrypting,
+strip the extra bytes again. Please note that this method will *not* work with all ciphers (e.g. XTS), it is
+something specific to CTR.
+
+#### Deriving the keys using HKDF
+As key deriving is needed for both encrypting and decrypting it makes sense to implement this first.
```

Comment:
Maybe worth noting that implementing this is only needed if you don't have access to an HKDF implementation.